### PR TITLE
Multivalue Sparse Vector Support

### PR DESCRIPTION
**Pull request by john-wagster**
Updated LuceneDocument to take advantage of looking up feature value on existing features and selecting the max when parsing multi-value sparse vectors

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/index/mapper/vectors/SparseVectorFieldMapperTests.java
Comment (kderusso) [inline server/src/test/java/org/elasticsearch/index/mapper/vectors/SparseVectorFieldMapperTests.java:153]: Nice test comments 😄

**Review by kderusso — COMMENTED** [review]
Looks great! Is it possible to test this behavior via a [yaml](https://github.com/elastic/elasticsearch/blob/main/rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/README.asciidoc) test?

**Review by benwtrent — COMMENTED** [review]
Need some yaml tests and documentation explaining that multi-value for spared vector will take the maximum dimension value. Maybe link the relevant literature as well?

**Comment by mayya-sharipova** [comment]
@elasticmachine generate changelog

**Comment by azasypkin** [comment]
Hey @john-wagster, I see you tagged the @elastic/kibana-security team for review, but I'm not sure there is anything in this PR that Kibana Security can help with. Maybe you meant to tag @elastic/es-security instead?

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/90_sparse_vector.yml
Comment (benwtrent) [inline rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/90_sparse_vector.yml:5]: we should instead add a new "Indexing and searching multi-valued sparse vectors in >=8.15"

And ensure this test still covers the single sparse vector test case in the appropriate versions.

File: server/src/main/java/org/elasticsearch/index/mapper/LuceneDocument.java
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/index/mapper/LuceneDocument.java:126]: What is the performance impact here? Its common to have 100s of fields, and those could have 100s of values. To my knowledge, `remove` does a linear scan of the values calling `equals` on each one. This ultimately just does an object identity check (as nobody overrides the `equals` method). 

One thought I had was to have the `keyedFields` actually be a map of
```
Map<Object, IndexableFieldAndPos> keyedFields
...
private record IndexableFieldAndPos(IndexableField value, int pos){}
```

This way we keep track of the position in the indexable list. This keeps us from doing two things:

 - Iterating the entire list on remove
 - Accidentally re-ordering the list (if somebody else uses this method in the future).

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/SparseVectorFieldMapper.java
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/index/mapper/vectors/SparseVectorFieldMapper.java:192]: 👍

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/SparseVectorFieldMapper.java
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/index/mapper/vectors/SparseVectorFieldMapper.java:198]: I am not sure we want to create a new field on every multi-value. 

Could we adjust the function definition to only be `calculateWithKey`, and it will pass `null` if it doesn't already exist. At that point a new one will be created and used, otherwise, we can call `setFeatureValue` on the existing one.

File: server/src/main/java/org/elasticsearch/index/mapper/LuceneDocument.java
Comment (john-wagster) [reply server/src/main/java/org/elasticsearch/index/mapper/LuceneDocument.java:126]: I'm glad you brought it up; I had some local notes to discuss with you on this.  Seemed like there were a couple of structures that could be updated either keyedFIelds as you suggest or the fields list could changed to be a Map like this this: 
`Map<Integer, IndexableField> fields` 
where the Integer here is a hashcode.  

Your suggestion is straightforward though I'll roll with that for now.

There's some weirdness in this class in general related to getting fields with the same name that I don't fully understand which might be worth further review.

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/90_sparse_vector.yml
Comment (john-wagster) [reply rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/90_sparse_vector.yml:5]: Ah that makes some sense; i'll fix this.

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/SparseVectorFieldMapper.java
Comment (john-wagster) [reply server/src/main/java/org/elasticsearch/index/mapper/vectors/SparseVectorFieldMapper.java]: Take a look at what I've got now (can always roll back).  But utilizing `setFeatureValue` (which I didn't realize was a good option here) let me drastically simplify the code.  LuceneDocument no longer has changes, no removal of features at all, and the logic can all sit sin SparseVectorFieldMapper.  

I may have missed some nuance of your suggestion here or some obvious requirement.  But as long as `setFeatureValue` has no detrimental downstream affects then it seems like a good solution?  

The concerns I had brought up in separate comment are related to how getting fields will return the first field by name but all of the sparse vector fields will have the same name.  Not sure if this is worth addressing here (vs creating a ticket) vs irrelevant but didn't want to lose that conversation.  We can discuss separately.

**Review by benwtrent — COMMENTED** [review]
Oh man! I love how simple it turned out to be :D!

Could you add some docs? The documentation would belong in `sparse-vector.asciidoc`

There is a note in that doc now that says:
```
NOTE: `sparse_vector` fields only support single-valued fields and strictly positive
values. Multi-valued fields and negative values will be rejected.
```

With this change, that would not be true. 

There should be a new section explaining multi value, headed with something like:

```
[[index-multi-value-sparse-vectors]]
==== Multi-value sparse vectors
```

We should explain that its a `max` value, and why we chose that and give an example.

**Comment by john-wagster** [comment]
@benwtrent I updated the docs but definitely want some feedback on if it's too much information or just off base.  My usage and understanding of sparse vectors or a decent example here might be wrong.

**Review by benwtrent — COMMENTED** [review]
This is good. I like it. 

Let's hold off on merging into lucene_snapshot. 

We can either reopen the PR against https://github.com/elastic/elasticsearch/pull/109219 

Or just wait for that PR to be merged, and rebase from main and merge the PR there :).

**Comment by benwtrent** [comment]
@tveasey mind taking a quick look? The logic is pretty simple (max dim), mainly I was concerned about the docs, that we got the gist of the idea correct.

File: docs/reference/mapping/types/sparse-vector.asciidoc
Comment (mayya-sharipova) [inline docs/reference/mapping/types/sparse-vector.asciidoc:43]: It would be nice to provide an explanation how we got `1.2`?

File: docs/reference/mapping/types/sparse-vector.asciidoc
Comment (mayya-sharipova) [inline docs/reference/mapping/types/sparse-vector.asciidoc:92]: Nit: `negative` -> `Negative`

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/90_sparse_vector.yml
Comment (mayya-sharipova) [inline rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/90_sparse_vector.yml:4]: What is the reason we are skipping this test for versions >= 8.15?

**Review by mayya-sharipova — APPROVED** [review]
@john-wagster Great work! Thanks John!

File: docs/reference/mapping/types/sparse-vector.asciidoc
Comment (john-wagster) [reply docs/reference/mapping/types/sparse-vector.asciidoc:92]: love nits; always welcome

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/SparseVectorFieldMapper.java
Comment (tveasey) [reply server/src/main/java/org/elasticsearch/index/mapper/vectors/SparseVectorFieldMapper.java:193]: So the best way to score long docs is to take the max over scores.

Suppose you have a document with passages `p_1`, `p_2`, ..., `p_n` and query `q` then you would score the document for the query as
```
score_best = max_i(q^t p_i) = max_i(sum_j(w_j(q)) x w_j(p_i)))
```
The second best way to score the document is to use max pooling of weights
```
score_runnerup = sum_j(w_j(q) x max_i(w_j(p_i)))
```
where `w_j` is the function returning the j'th token weight. I may be missing the context, but my understanding here is that feature refers to `w_j` so we use the second approach. This is ok, but it tends to suffer more on docs with many passages than the first approach.

All that said, if you are indexing the joint representation, i.e. single sparse vector with max token weights to provide a tradeoff for index size and quality then this is absolutely fine. My comment only applies if this is the only way we are considering scoring long docs from their passages.

**Review by tveasey — COMMENTED** [review]
I just had one clarifying comment / question.

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/90_sparse_vector.yml
Comment (mayya-sharipova) [reply rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/90_sparse_vector.yml:4]: Makes sense to skip the repetition!

**Comment by mayya-sharipova** [comment]
@tveasey Yes, exactly, it is the 2nd approach.

For the 1st approach could be implemented by using sparse_vectors as nested fields. In this case a document will be scored by its best matching paragraph.

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/SparseVectorFieldMapper.java
Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/index/mapper/vectors/SparseVectorFieldMapper.java:193]: @tveasey 

> My comment only applies if this is the only way we are considering scoring long docs from their passages.

Definitely not. Folks can still use `nested` if they so wish.

File: docs/reference/mapping/types/sparse-vector.asciidoc
Comment (benwtrent) [inline docs/reference/mapping/types/sparse-vector.asciidoc:43]:
```suggestion
More specifically the final `GET` query here returns a `_score` of ~1.2 (which is the `max(impact.delicious[0], impact.delicious[1])` and is approximate because we have a relative error of 0.4% as explained below)
```

**Review by benwtrent — APPROVED** [review]

**Pull request merged**
